### PR TITLE
Fix unreachable sub-branch detection in or-patterns

### LINKED ISSUE #76836: unreachable_patterns false negative with or_patterns
The match remains exhaustive with highlighted pattern removed, but it does not produce any warnings:

```rust
#![feature(or_patterns)]

pub struct A;

pub enum B {
    X,
    Y,
}

pub enum E {
    A(A),
    B(B),
}

pub fn f(e: E) -> () {
    match e {
        E::B(B::X) => (),
        E::B(B::Y | B::X)
        //          ~~~~ 
        | E::A(A) => (),
    }
}
```

### PR DESCRIPTION
**Pull request by Nadrieril**
The previous implementation was too eager to avoid unnecessary "unreachable pattern" warnings. I feel more confident about this implementation than I felt about the previous one.
Fixes https://github.com/rust-lang/rust/issues/76836.

@rustbot modify labels: +A-exhaustiveness-checking

### CODE REVIEW

File: src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr [hunk 1]
```diff
@@ -77,15 +77,15 @@ LL |         (1 | 1,) => {}
    |              ^
 
 error: unreachable pattern
-  --> $DIR/exhaustiveness-unreachable-pattern.rs:53:15
+  --> $DIR/exhaustiveness-unreachable-pattern.rs:55:15
    |
-LL |             | 0
+LL |             | 0] => {}
```
Comment (varkor) [inline src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr:82]: I'm not sure how much we care about emitting match errors in the order in which they occur, but if it's easy to make prior patterns warn earlier, that's probably slightly better for the user.

Comment (Nadrieril) [reply src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr]: It's quite easy yes, it appears `Span`s are `Ord` wrt their position in a file.

**Comment by varkor** [comment]
Thanks!

@bors r+ rollup

**Pull request merged**